src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -40,91 +40,107 @@
+    );
+
+    request.setAttribute("recipesList", recipesListToReturn);
+    request.getRequestDispatcher("/search.jsp").forward(request, response);
+  }
+
+  /**
+   * Returns a list of recipes matching the given title and ingredients.
+   */
+  private List<Recipe> recipesMatching(
+    HttpServletRequest request,
+    HttpServletResponse response,
+    String stringToMatch
+  )
+    throws ServletException, IOException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Recipe> recipesMatchingList = new ArrayList<>();
+    Index index = getIndex("recipes_index");
 
-    Query query = buildQuery(titleToMatch);
-    Results<ScoredDocument> resultsByTitle = searchIndex.search(query);
-    for (ScoredDocument entity : resultsByTitle) {
+    Query query = buildQuery(stringToMatch);
+    Results<ScoredDocument> results = index.search(query);
+    for (ScoredDocument entity : results) {
       try {
         Entity recipeEntity = getRecipeEntityFromDocumentEntity(
           datastore,
           entity
         );
+
         Recipe recipe = buildRecipe(recipeEntity);
-        recipesList.add(recipe);
+        recipesMatchingList.add(recipe);
       } catch (EntityNotFoundException e) {
         response.setStatus(505);
         request.getRequestDispatcher("/search.jsp").forward(request, response);
-        return;
+        return recipesMatchingList;
       }
     }
-    request.setAttribute("recipesList", recipesList);
-    request.getRequestDispatcher("/search.jsp").forward(request, response);
+    return recipesMatchingList;
   }
 
   /**
    * Returns the index that stores the recipes documents
    */
-  private Index getIndex() {
-    IndexSpec indexSpec = IndexSpec
-      .newBuilder()
-      .setName("recipesIndex")
-      .build();
+  private Index getIndex(String indexName) {
+    IndexSpec indexSpec = IndexSpec.newBuilder().setName(indexName).build();
     Index index = SearchServiceFactory.getSearchService().getIndex(indexSpec);
     return index;
   }
 
-  private Query buildQuery(String titleToMatch) {
+  private Query buildQuery(String stringToMatch) {
     Cursor responseCursor = Cursor.newBuilder().build();
     QueryOptions options = QueryOptions
       .newBuilder()
-      .setLimit(10)
-      .setFieldsToSnippet("title")
-      .setCursor(responseCursor)
+      .setLimit(60)
       .setSortOptions(
-        SortOptions
-          .newBuilder()
-          .addSortExpression(
-            SortExpression
-              .newBuilder()
-              .setExpression("title")
-              .setDirection(SortExpression.SortDirection.ASCENDING)
-              .setDefaultValue("")
-          )
+        SortOptions.newBuilder().setMatchScorer(MatchScorer.newBuilder())
       )
+      .setCursor(responseCursor)
       .build();
 
     Query query = Query
       .newBuilder()
       .setOptions(options)
-      .build(titleToMatch.toLowerCase());
+      .build(stringToMatch.toLowerCase());
```

Comment:
Are you sure the toLowerCase is needed? In the documentation this says that stemming is done on the search parameters, so it shouldn't be needed :).

Reply:
No, it is not needed. I removed it.